### PR TITLE
[@shopify/react-server-webpack-plugin] Add support for node projects

### PR DESCRIPTION
**Pull request by ismail-syed**
## Description

Fixes https://github.com/Shopify/quilt/issues/903

This PR adds Node project support to `@shopify/react-server-webpack-plugin` with respect to the project layout [sewing-kit](https://github.com/shopify/sewing-kit#project-layout-node) expects. 

This is done by: 
* Adding a `rails` option to the webpack plugin to let it know whether we're virtualizing modules for a Rails or Node project. 
* Making it aware of an `/app/index.js` file for Node projects. Creates a virtual module at `./app/index.js` when appropriate

### 🎩 

- `cd packages/react-server-webpack-plugin && yarn run build && yarn link`
- `dev cd sewing-kit` and checkout `react-server-webpack-plugin-node-support`
  - `yarn link @shopify/react-server-webpack-plugin && yarn run build`
  - `yarn link @shopify/sewing-kit`
-  `dev cd web-proving-ground && yarn link @shopify/sewing-kit`
  - `yarn dev` and verify that the project actually runs
  - Stop the server. Delete the `client` directory. Run the server & verify that the app still works
  - Stop the server. Delete the `server` directory. Run the server & verify that the app still works
  - Stop the server. Delete the `app` directory. Run the server & verify that the app still works. At this point the whole app is pretty much running off virtualized webpack modules. You'll see a beautiful boilerplate component being rendered. Magic 😄 

Screen Shot 2019-08-27 at 2 59 45 PM

-  `dev cd web-rails-proving-ground && yarn link @shopify/sewing-kit`
  - Verify that the existing behavior for rails apps still works.

## Type of change

- [x]  Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above

### CODE REVIEW

File: packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts [hunk 1]
```diff
@@ -43,70 +55,113 @@ export class ReactServerPlugin {
...
+    {inputFileSystem, options: {context = ''}}: Compiler,
+  ) {
+    const {basePath: path, rails} = this.options;
+    const basePath = resolve(context, path);
+    const resolvedPath = rails ? `${basePath}` : `${basePath}/${entry}`;
+
+    if (!rails) {
+      if (!existsSync(resolvedPath)) {
+        // In node projects, if /client, /server or /app dirs dont exist, create them so VirtualModulesPlugin can resolve the modules paths
+        mkdir(resolvedPath);
```
[the remark this answers is not in the document]
Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts]: PR has be updated, no more ugly `mkdir`

File: packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts [hunk 2]
```diff
@@ -1,112 +1,162 @@
...
 
   apply(compiler: Compiler) {
     const modules = this.modules(compiler);
-    const virtualModules = (new VirtualModulesPlugin(modules) as any) as Plugin;
-    virtualModules.apply(compiler);
+    const virtualModules = new VirtualModulesPlugin(modules) as any;
+    (virtualModules as Plugin).apply(compiler);
+
+    compiler.hooks.compilation.tap('ReactServerPlugin', compilation => {
+      if (this.error) {
```
[the remark this answers is not in the document]
Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts]: Yes, it seems it doesn't work in watch mode. What hook compiler hook runs on changes, I tried our `compiler.hook.watchRun` but no dice :

**Comment by ismail-syed** [comment]
All the comments have been addressed except the one about [watch mode](https://github.com/Shopify/quilt/pull/917#discussion_r319145326). Will tackle that one tomorrow morning. Aside from that this is in a good state for another review. 

Updates: 

- app imports are now all pulling from `index` since we the app path resolved for us by sewing-kit. Test have been updated with `resolves.modules` to reflect this. Tested in both web-proving-ground and web-rails-proving-ground
-  Chatted with Gord about this earlier today. In node, if a user has an existing `app`/`client`/`server` directory, we can assume we know what they're doing. No need to create the folder and cover those wonky edge cases. 
- Cleaned up the unnecessary private functions
-  Removed the use of `inputFileSystem` in favor of `glob`/`fs`
- Cleaned up the webpack configs in tests to return with the suggestion of returning a function instead of the function.

File: packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts [hunk 3]
```diff
@@ -27,85 +34,102 @@ export class ReactServerPlugin {
...
-          import {showPage} from '@shopify/react-html';
-          import App from './index';
-
-          const appContainer = document.getElementById('app');
-
-          ReactDOM.hydrate(React.createElement(App), appContainer);
-          showPage();
-        `;
+    if (noSourceExists(Entrypoint.Client, this.options, compiler)) {
+      const file = `${join(basePath, Entrypoint.Client)}.js`;
```
Comment (lemonmade) [inline packages/react-server-webpack-plugin/src/react-server-webpack-plugin.ts]: Put the extension only on the filename (`${Entrypoint.Client}.js`)

File: packages/react-server-webpack-plugin/src/test/react-server-webpack-plugin.test.ts [hunk 4]
```diff
@@ -1,65 +1,71 @@
...
+        const clientModule = clientResults.modules.find(({name}) =>
+          name.includes('./client.js'),
+        );
+
+        const serverModule = serverResults.modules.find(({name}) =>
+          name.includes('./server.js'),
+        );
+
+        expect(clientModule.source).toMatch(HEADER);
+        expect(serverModule.source).toMatch(HEADER);
```
Comment (lemonmade) [inline packages/react-server-webpack-plugin/src/test/react-server-webpack-plugin.test.ts]: Do we actually care about the header?

Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/test/react-server-webpack-plugin.test.ts]: It's how we verify that the plugin ran. What alternatives did you have in mind?

Comment (lemonmade) [reply packages/react-server-webpack-plugin/src/test/react-server-webpack-plugin.test.ts]: Wouldn't any contents prove that? In the end, a user doesn't care about any of these things, so I'd rather see tests that actually show you how a user would experience the plugin, but at least let's make it more clear why this assertion means things are working.

File: packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js [hunk 5]
```diff
@@ -1,33 +1,39 @@
...
+  const universal = {
+    mode: 'production',
+    optimization: {
+      minimize: false,
     },
-  ],
-};
+    plugins: [new ReactServerPlugin()],
+    resolve: {
+      modules: ['node_modules', path.resolve(__dirname)],
```
Comment (GoodForOneFare) [inline packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js]: Will `path.resolve` be `app/ui` for Rails tests?

Comment (GoodForOneFare) [reply packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js]: I don't think it will, and I think it's a smell that the test setup is too oblique.  Consider just creating `test/fixtures/rails` and `test/fixtures/node` and duplicating setups.

File: packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
-  name: 'server',
-  target: 'node',
-  entry: './server',
-  externals: [
-    (context, request, callback) => {
-      if (/node_modules/.test(context)) {
-        return callback(null, `commonjs ${request}`);
-      }
-      callback();
+module.exports = function config(usesNode) {
```
[the remark this answers is not in the document]
Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js]: Not needed, anymore. Good find.

File: packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js (hunk 5 above)
Comment (ismail-syed) [reply packages/react-server-webpack-plugin/src/test/fixtures/no-entrypoints/webpack.config.js]: > Will path.resolve be app/ui for Rails tests?

Not necessarily. The need for `app/ui` is based on the assumption that a project is using sewing-kit rails setup, which these tests aren't.

This config adds the root path to the module resolution so `import App from 'index';` resovles to `__dirname/index.js`

**Comment by ismail-syed** [comment]
Got an offline 👌 from Gord, merging.
